File: ts_to_heifer/test/heifer_output/04_operators.ml
=== Translation Successful ===

Heifer IR:
let add_nums = fun x y (*@ req x:int/\y:int; ens res:int @*) -> (x + y) in
let subtract = fun x y (*@ req x:int/\y:int; ens res:int @*) -> (x - y) in
let multiply = fun x y (*@ req x:int/\y:int; ens res:int @*) -> (x * y) in
let divide = fun x y (*@ req x:int/\y:int; ens res:int @*) -> (x / y) in
let power = fun x y (*@ req x:int/\y:int; ens res:int @*) -> (x ^ y) in
let less_than = fun x y (*@ req x:int/\y:int; ens res:bool @*) -> if x<y then true else (false) in
let greater_or_equal = fun x y (*@ req x:int/\y:int; ens res:bool @*) -> if x>=y then true else (false) in
let equals = fun x y (*@ req x:int/\y:int; ens res:bool @*) -> if x=y then true else (false) in
let not_equals = fun x y (*@ req x:int/\y:int; ens res:bool @*) -> if not(x=y) then true else (false) in
let and_op = fun a b (*@ req a:bool/\b:bool; ens res:bool @*) -> (a && b) in
let or_op = fun a b (*@ req a:bool/\b:bool; ens res:bool @*) -> (a || b) in
let not_op = fun a (*@ req a:bool; ens res:bool @*) -> true in
let concat = fun a b (*@ req a:Str/\b:Str; ens res:Str @*) -> (a + b) in
let negate = fun x (*@ req x:int; ens res:int @*) -> (0 - x) in
let complex_expr = fun x y z (*@ req x:int/\y:int/\z:int; ens res:int @*) -> (((x + y) * z) - ((x - y) / 2)) in
()

Type: unit

✓ Output is in Heifer's native format

File: ts_to_heifer/test/test_cases/ts/04_operators.ts
// Test 4: Binary and unary operators
// Testing operator parsing and types

// Arithmetic operators
function add_nums(x: number, y: number): number {
  return x + y;
}

function subtract(x: number, y: number): number {
  return x - y;
}

function multiply(x: number, y: number): number {
  return x * y;
}

function divide(x: number, y: number): number {
  return x / y;
}

function power(x: number, y: number): number {
  return x ** y;
}

// Comparison operators
function less_than(x: number, y: number): boolean {
  return x < y;
}

function greater_or_equal(x: number, y: number): boolean {
  return x >= y;
}

function equals(x: number, y: number): boolean {
  return x === y;
}

function not_equals(x: number, y: number): boolean {
  return x !== y;
}

// Logical operators
function and_op(a: boolean, b: boolean): boolean {
  return a && b;
}

function or_op(a: boolean, b: boolean): boolean {
  return a || b;
}

function not_op(a: boolean): boolean {
  return !a;
}

// String concatenation
function concat(a: string, b: string): string {
  return a + b;
}

// Unary minus
function negate(x: number): number {
  return -x;
}

// Complex expression
function complex_expr(x: number, y: number, z: number): number {
  return (x + y) * z - (x - y) / 2;
}
